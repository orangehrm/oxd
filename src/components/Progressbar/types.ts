/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */

export const TYPE_DEFAULT = 'default';
export const TYPE_MAIN = 'main';
export const TYPE_SECONDARY = 'secondary';
export const TYPE_SUCCESS = 'success';
export const TYPE_WARN = 'warn';
export const TYPE_ERROR = 'error';
export const TYPE_INFO = 'info';

export const TYPES = [
  TYPE_DEFAULT,
  TYPE_MAIN,
  TYPE_SECONDARY,
  TYPE_SUCCESS,
  TYPE_WARN,
  TYPE_ERROR,
  TYPE_INFO,
];

export type ProgressbarType =
  | typeof TYPE_DEFAULT
  | typeof TYPE_MAIN
  | typeof TYPE_SECONDARY
  | typeof TYPE_SUCCESS
  | typeof TYPE_WARN
  | typeof TYPE_ERROR
  | typeof TYPE_INFO;
