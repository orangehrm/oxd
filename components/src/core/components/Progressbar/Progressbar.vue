<!--
/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */
-->

<template>
  <div class="oxd-progressbar">
    <div class="oxd-progressbar-wrapper">
      <div :style="styles" :class="classes"></div>
    </div>
    <oxd-text v-if="showLabel" class="oxd-progressbar-text" tag="p">
      {{ progressText }}
    </oxd-text>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import {ProgressbarType, TYPE_DEFAULT, TYPES} from './types';
import Text from '@ohrm/oxd/core/components/Text/Text.vue';

export default defineComponent({
  name: 'oxd-progressbar',
  props: {
    showLabel: {
      type: Boolean,
      default: true,
    },
    stripe: {
      type: Boolean,
      default: false,
    },
    animation: {
      type: Boolean,
      default: false,
    },
    progress: {
      type: Number,
      default: 0,
      validator: function(value: number) {
        return value >= 0 && value <= 100;
      },
    },
    type: {
      type: String,
      default: TYPE_DEFAULT,
      validator: function(value: ProgressbarType) {
        return TYPES.indexOf(value) !== -1;
      },
    },
  },
  components: {
    'oxd-text': Text,
  },
  computed: {
    styles(): object {
      return {
        width: `${this.progress}%`,
      };
    },
    classes(): object {
      return {
        'oxd-progressbar-ticker': true,
        [`oxd-progressbar-ticker--${this.type}`]: true,
        '--striped': this.stripe,
        '--animated': this.animation,
      };
    },
    progressText(): string {
      return `${Math.round(this.progress)}%`;
    },
  },
});
</script>

<style src="./progressbar.scss" lang="scss" scoped></style>
