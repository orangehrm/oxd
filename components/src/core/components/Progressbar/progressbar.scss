// This file is part of OrangeHRM Inc
//
// Copyright (C) 2020 onwards OrangeHRM Inc
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see  http://www.gnu.org/licenses

@import '../../../styles';
@import 'variables';

.oxd-progressbar {
  display: flex;
  align-items: center;
  &-text {
    white-space: nowrap;
    margin-left: $oxd-progressbar-text-margin;
    font-size: $oxd-progressbar-font-size;
    font-weight: $oxd-progressbar-font-weight;
  }
  &-wrapper {
    width: 100%;
    overflow: hidden;
    height: $oxd-progressbar-height;
    border-radius: $oxd-progressbar-border-radius;
    background-color: $oxd-progressbar-background-color;
  }
  &-ticker {
    height: $oxd-progressbar-height;
    border-radius: $oxd-progressbar-border-radius;
    background-color: $oxd-progressbar-color-default;

    &--main {
      background-color: $oxd-progressbar-color-main;
    }

    &--secondary {
      background-color: $oxd-progressbar-color-secondary;
    }

    &--success {
      background-color: $oxd-progressbar-color-success;
    }

    &--warn {
      background-color: $oxd-progressbar-color-warn;
    }

    &--error {
      background-color: $oxd-progressbar-color-error;
    }

    &--info {
      background-color: $oxd-progressbar-color-info;
    }

    &.--striped {
      background-image: $oxd-progressbar-bg-striped;
      background-size: $oxd-progressbar-bg-size;
    }
    &.--animated {
      animation: $oxd-progressbar-animation;
    }
  }
}

@keyframes move {
  0% {
    background-position: 0 0;
  }
  100% {
    background-position: $oxd-progressbar-bg-size;
  }
}
