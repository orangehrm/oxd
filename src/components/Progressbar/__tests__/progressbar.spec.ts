/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */

import {mount} from '@vue/test-utils';
import {describe, expect, it} from 'vitest';
import Progressbar from '@/components/Progressbar/Progressbar.vue';

describe('Progressbar.vue', () => {
  it('renders OXD Progressbar', () => {
    const wrapper = mount(Progressbar, {
      props: {},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('shows correct progress in Progressbar', () => {
    const wrapper = mount(Progressbar, {
      props: {
        progress: 25,
      },
    });
    expect(wrapper.find('.oxd-progressbar-ticker').attributes().style).toBe(
      'width: 25%;',
    );
  });
});
