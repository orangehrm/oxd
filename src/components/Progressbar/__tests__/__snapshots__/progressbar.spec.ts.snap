// Vitest Snapshot v1

exports[`Progressbar.vue > renders OXD Progressbar 1`] = `
"<div class=\\"oxd-progressbar\\" data-v-668268dd=\\"\\">
  <div class=\\"oxd-progressbar-wrapper\\" data-v-668268dd=\\"\\">
    <div style=\\"width: 0%;\\" class=\\"oxd-progressbar-ticker oxd-progressbar-ticker--default\\" data-v-668268dd=\\"\\"></div>
  </div>
  <p class=\\"oxd-text oxd-text--p oxd-progressbar-text\\" data-v-0334bd57=\\"\\" data-v-668268dd=\\"\\">0%</p>
</div>"
`;
